src/java/org/apache/cassandra/repair/consistent/CoordinatorSession.java

```
@@ -287,103 +280,55 @@ private static String formatDuration(long then, long now)
-            {
-                repairStart = currentTimeMillis();
-                if (logger.isDebugEnabled())
-                {
-                    logger.debug("Incremental repair {} prepare phase completed in {}", sessionID, formatDuration(sessionStart, repairStart));
-                }
-                setRepairing();
-                return sessionSubmitter.get();
-            }
-            else
-            {
-                return ImmediateFuture.success(null);
-            }
+        Future<CoordinatedRepairResult> repairSessionResults = prepareResult.flatMap(ignore -> {
+            repairStart = currentTimeMillis();
+            if (logger.isDebugEnabled())
+                logger.debug("Incremental repair {} prepare phase completed in {}", sessionID, formatDuration(sessionStart, repairStart));
+            setRepairing();
+            return sessionSubmitter.get();
         });
 
-        // mark propose finalization
-        Future<Boolean> proposeFuture = repairSessionResults.andThenAsync(results ->
-        {
-            if (results == null || results.isEmpty() || Iterables.any(results, r -> r == null))
+        // if any session failed, then fail the future
+        Future<CoordinatedRepairResult> onlySuccessSessionResults = repairSessionResults.map(result -> {
+            if (result.isFailed())
             {
+                //TODO this looks like a bug... why does this only update IFF repair failed?
                 finalizeStart = currentTimeMillis();
                 if (logger.isDebugEnabled())
-                {
                     logger.debug("Incremental repair {} validation/stream phase completed in {}", sessionID, formatDuration(repairStart, finalizeStart));
-                }
-                return ImmediateFuture.failure(SomeRepairFailedException.INSTANCE);
-            }
-            else
-            {
-                return finalizePropose();
+                throw SomeRepairFailedException.INSTANCE;
```

since this is a cached exception, the cost to throw mostly matches return, and the Future interface makes sure we properly convert exceptions into failed futures; so don't really need to return `ImmediateFuture.failure(SomeRepairFailedException.INSTANCE)`; so both are basically the same

---

Also, IR is different than normal/preview as it fails the future rather than returning failed results.  At the top level they get translated to the same thing

```
f.addCallback((result, failure) -> {
            try
            {
                if (failure != null)
                {
                    notifyError(failure);
                    fail(failure.getMessage());
                }
                else
                {
                    maybeStoreParentRepairSuccess(result.successfulRanges);
                    if (result.isFailed())
                    {
                        fail(null);
                    }
                    ...
```
                    
`notifyError` ignores `SomeRepairFailedException`, so acts as a no-op
`fail` will see a null as there is no message

The main difference will be the lack of `maybeStoreParentRepairSuccess(result.successfulRanges)`

---

I could fix this difference in this PR, but figured the current patch was large enough so left as-is; let me know if you feel we should standardize this, wouldn't be that big a deal as we could handle that at `onlySuccessSessionResults.flatMap` by checking then no-oping.

---

I think if we'll be throwing this we should disable suppressed exceptions and writable stack traces in the constructor of the static instance.

---

good advice, updated it

```
private SomeRepairFailedException()
    {
        super(null, null, false, false);
    }
```